### PR TITLE
sysupdate: Handle incomplete versions

### LINKED ISSUE #33339: Handling of incomplete sysupdate components
### systemd version the issue has been seen with

256

### Used distribution

GNOME OS

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

When adding a new transfer to `sysupdate.d`, systemd-sysupdate should download the newly added files (e.g. sysexts) while protecting the currently booted slot.

### Unexpected behaviour you saw

I'm seeing two incorrect behaviours, but I think they are linked so filing in one issue.

1. If a new transfer is added in `sysupdate.d`, the currently installed version is now an incomplete version of what is supposed to be installed based on the configuration. sysupdate currently considers it to be not installed, and so doesn't protect it.

2. When already using the version we're updating to (e.g. already on the latest version, or explicitly requesting the current version) sysupdate downloads it into the other slot rather than complete the current incomplete version.

### Steps to reproduce the problem

1. Update to the latest version and reboot
2. Copy additional config files from `/usr/lib/sysregistry.d/` to `/etc/sysupdate.d`
3. Check what version is installed (`updatectl list .host`). This shows no version installed.
4. Update the system (`updatectl update`)

(We're currently carrying the patches from https://github.com/systemd/systemd/pull/32363 so you can use `updatectl`, but the bug can be reproduced by calling `systemd-sysupdate` directly)

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by AdrianVovk**
An incomplete version is one that is partially installed/available: some transfers of this version exist, but others don't.

Incomplete versions coming from the remote are completely ignored: if some transfer is missing from a given version, we just pretend that the version outright doesn't exist.

Incomplete versions that are installed are repaired in place. This is a situation where the user might have deleted a file that should be installed by a transfer, or eventually it will be used to support "optional features" that dynamically add/remove transfers from the installation. 

Depends on:
- #33520
- #32363
- #34079

Fixes #33339

### CODE REVIEW

File: src/sysupdate/sysupdate.c [hunk 1]
```diff
@@ -670,6 +700,7 @@ static int context_show_version(Context *c, const char *version) {
                                           SD_JSON_BUILD_PAIR_BOOLEAN("installed", FLAGS_SET(us->flags, UPDATE_INSTALLED)),
                                           SD_JSON_BUILD_PAIR_BOOLEAN("obsolete", FLAGS_SET(us->flags, UPDATE_OBSOLETE)),
                                           SD_JSON_BUILD_PAIR_BOOLEAN("protected", FLAGS_SET(us->flags, UPDATE_PROTECTED)),
+                                          SD_JSON_BUILD_PAIR_BOOLEAN("incomplete", FLAGS_SET(us->flags, UPDATE_INCOMPLETE)),
```
Comment (abderrahim) [inline src/sysupdate/sysupdate.c:734]: updatectl needs to be updated to handle this. Something like

```diff
diff --git a/src/sysupdate/updatectl.c b/src/sysupdate/updatectl.c
index 0ef61bf895..49cb375720 100644
--- a/src/sysupdate/updatectl.c
+++ b/src/sysupdate/updatectl.c
@@ -360,6 +360,10 @@ static int parse_describe(sd_bus_message *reply, Version *ret) {
                         assert(sd_json_variant_is_boolean(value));
                         if (sd_json_variant_boolean(value))
                                 v.flags |= UPDATE_PROTECTED;
+                } else if (streq(key, "incomplete")) {
+                        assert(sd_json_variant_is_boolean(value));
+                        if (sd_json_variant_boolean(value))
+                                v.flags |= UPDATE_INCOMPLETE;
                 } else if (streq(key, "changelog-url")) {
                         sd_json_variant *entry;
                         assert(sd_json_variant_is_array(value));
```

**Comment by abderrahim** [comment]
Tested this MR and it works fine (with the above patch to updatectl)

File: src/sysupdate/sysupdate.c (hunk 1 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdate.c:734]: Good point. I'll address this & rebase once #32363 lands

Comment (AdrianVovk) [reply src/sysupdate/sysupdate.c:734]: Need to make sure it's documented (and supported?) in sysupdated too.

File: src/sysupdate/sysupdate-resource.c [hunk 2]
```diff
@@ -527,13 +527,25 @@ int resource_load_instances(Resource *rr, bool verify, Hashmap **web_cache) {
         return 0;
 }
 
+static int instance_version_match(Instance *const*a, Instance *const*b) {
+        assert(a);
+        assert(b);
+        assert(*a);
+        assert(*b);
+        assert((*a)->metadata.version);
```
Comment (bluca) [inline src/sysupdate/sysupdate-resource.c:535]: This will abort if a version is `0`, is that intended?

**Review by bluca — APPROVED** [review]

File: src/sysupdate/sysupdate-resource.c (hunk 2 above)
Comment (AdrianVovk) [reply src/sysupdate/sysupdate-resource.c:535]: the version is a char*, not a number

**Comment by AdrianVovk** [comment]
Test failure seems unrelated:

```
11001s Summary of Failures:
11001s 
11001s  1/92 systemd:integration-tests / TEST-64-UDEV-STORAGE-btrfs_basic                              FAIL              9.68s   (exit status 247 or signal 119 SIGinvalid)
11001s  2/92 systemd:integration-tests / TEST-64-UDEV-STORAGE-iscsi_lvm                                FAIL             14.69s   (exit status 247 or signal 119 SIGinvalid)
11001s 11/92 systemd:integration-tests / TEST-64-UDEV-STORAGE-virtio_scsi_basic                        FAIL              6.97s   (exit status 247 or signal 119 SIGinvalid)
11001s 
11001s Ok:                 87  
11001s Expected Fail:      0   
11001s Fail:               3   
11001s Unexpected Pass:    0   
11001s Skipped:            2   
11001s Timeout:            0   
```

**Pull request merged**